app/feature/feature-claim-details/src/main/kotlin/com/hedvig/android/feature/claim/details/ui/ClaimAsyncImage.kt

```diff
@@ -0,0 +1,69 @@
+package com.hedvig.android.feature.claim.details.ui
+
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.unit.IntSize
+import coil.ImageLoader
+import coil.compose.AsyncImage
+import coil.compose.AsyncImagePainter
+import coil.request.ImageRequest
+import com.hedvig.android.core.designsystem.material3.rememberShapedColorPainter
+import com.hedvig.android.core.designsystem.material3.squircleMedium
+import com.hedvig.android.core.ui.layout.adjustSizeToImageRatio
+import com.hedvig.android.placeholder.PlaceholderHighlight
+import com.hedvig.android.placeholder.fade
+import com.hedvig.android.placeholder.placeholder
+
+@Composable
+fun ClaimAsyncImage(model: Any, imageLoader: ImageLoader, modifier: Modifier = Modifier, cacheKey: String? = null) {
+  val loadedImageIntrinsicSize = remember { mutableStateOf<IntSize?>(null) }
+  val placeholderPainter: Painter = rememberShapedColorPainter(MaterialTheme.colorScheme.surface)
+  AsyncImage(
+    model = ImageRequest.Builder(LocalContext.current)
+      .data(model)
+      .apply {
+        if (cacheKey != null) {
+          diskCacheKey(cacheKey).memoryCacheKey(cacheKey)
+        }
+      }
+      .build(),
+    contentDescription = null,
+    imageLoader = imageLoader,
+    transform = { state ->
+      when (state) {
+        is AsyncImagePainter.State.Loading -> {
+          state.copy(painter = placeholderPainter)
+        }
+
+        is AsyncImagePainter.State.Error -> {
+          state
+        }
+
+        AsyncImagePainter.State.Empty -> state
+        is AsyncImagePainter.State.Success -> {
+          loadedImageIntrinsicSize.value = IntSize(
+            state.result.drawable.intrinsicWidth,
+            state.result.drawable.intrinsicHeight,
+          )
+          state
+        }
+      }
+    },
+    modifier = modifier
+      .adjustSizeToImageRatio(getImageSize = { loadedImageIntrinsicSize.value })
```

Comment:
Yeah interesting idea to reuse this from chat.
I think that worked great there with the constraints that we had which was that it may take various widths too. In this scenario as far as I can tell we'll have a fixed width, since each item will be taking 1/3rd of the screen.

May I propose an alternative instead here. Could you look into using https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/lazy/staggeredgrid/LazyStaggeredGrid.kt;l=39?q=LazyStaggeredGrid&sq= instead?
And try not to give them any constraints vertically, as they will always have a fixed width, so they should be able to adjust their height as they wish. And using the Staggered grid instead will make the things fall into place naturally.

With all this said, I recognize that `adjustSizeToImageRatio` is not a very descriptive name, and maybe I should've picked a better name, but my idea was that it's only gonna be used in chat so I didn't need to make the name super big :D Hence it has some constants in there, especially around
```
private const val MaximumImageHeight = 250
private const val PreferredImageHeight = 180
private const val MinimumImageHeight = 100
private const val MinimumWidthTakenPercentage = 0.5f
```
if you see in there. Which describe that the images should be taking from 50% of the width to 100%. And with extra constraints of the heights being forced into 100dp..250dp.